Add support for multiple (named) return parameters in Natspec devdoc

corydickson opened the pull request:
### Description
Addresses #1141 

Adds an object similar to the one used to document `params` only if multiple named return parameters are specified. Otherwise the `return` field will contain a single string.

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] README / documentation was extended, if necessary(?)
- [x] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

corydickson committed:
Refactor to have multiple return params

gh1dra commented on libsolidity/interface/Natspec.cpp:
```diff
@@ -99,16 +99,46 @@ Json::Value Natspec::devDocumentation(ContractDefinition const& _contractDef)
+					}
+
+					else {
+            			value += i->second.content;
```
Right now this piece of code handles multiple **unnamed** return parameters by concatenating all the `return` doc notices it encounters like before.

gh1dra commented:
On my machine the gnosis tests are failing due to segfault; investigating still...

corydickson committed:
Add unamed return param test and check size in conditional

erak commented on test/libsolidity/SolidityNatspecJSON.cpp:
```diff
@@ -415,6 +415,72 @@ BOOST_AUTO_TEST_CASE(dev_return_desc_after_nl)
+	"            \"a\": \"Documentation for the first parameter starts here. Since it's a really complicated parameter we need 2 lines\",\n"
+	"            \"second\": \"Documentation for the second parameter\"\n"
+	"        },\n"
+	"        \"return\": \"The result of the multiplicationAnd cookies with nutella\"\n"
```
I think this needs to change as well such that `return` _always_ contains at least one key-value-pair for each return parameter. Since they don't have a name in this test, we could think about adding a unique placeholder for each parameter:
```
"return": {
    "_1": "The result of the multiplication",
    "_2": "And cookies with nutella"
},
```
If the parameter has a name it could use that as the key.

I think we can't check if the documentation contains the name of the parameter (as we do for function parameters) since they are allowed to have no name (the compiler allows function parameters to don't have a name, but NatSpec doesn't).

erak commented on test/libsolidity/SolidityNatspecJSON.cpp:
```diff
@@ -415,6 +415,72 @@ BOOST_AUTO_TEST_CASE(dev_return_desc_after_nl)
+			/// @return The result of the multiplication
+			/// @return And cookies with nutella
+			function mul(uint a, uint second) public returns (uint, uint) {
+        		uint mul = a * 7;
```
Indentation is off.

erak reviewed (changes requested):
I think we need to revisit the concept again. E.g. We should also take the first word of the line and check if this corresponds to named return parameter, if any. If return parameters don't have a name, this check should be ignored.

erak commented:
Please also add a Changelog entry since this is a user-facing change that will effect tools.

corydickson commented:
> I think we need to revisit the concept again. E.g. We should also take the first word of the line and check if this corresponds to named return parameter, if any. If return parameters don't have a name, this check should be ignored.

Would it make sense to have this method perform that check and throw an error? It might be cleaner to reuse the current logic in the doc parser for `param` adding the `return` case for `parseDocTag`.  Though this would break most contract documentation i.e. the gnosis contract examples, no?

erak commented:
@gh1dra Yes, I think we should use the same logic as we do for function parameters. And yes, you're right. This would break existing contracts that use named return parameters (and that probably do not use the parameter name as the first word). But since the current behaviour is broken, I'd tend to properly clean it up and include this PR in our next breaking release, which will be 0.6.0.

erak commented:
The failing macOS build (`b_osx`) should be gone after a rebase, which will include https://github.com/ethereum/solidity/pull/7552.

corydickson committed:
Give unamed parameters unique keys and update tests to new spec

Fix whitespace

corydickson commented on libsolidity/interface/Natspec.cpp:
```diff
@@ -100,15 +100,52 @@ Json::Value Natspec::devDocumentation(ContractDefinition const& _contractDef)
+							auto nameEndPos = content.find_first_of(" \t");
+							solAssert(content.substr(0, nameEndPos) == paramName, "No return param name given: " + paramName);
+							content = content.substr(nameEndPos+1);
+						}
```
@erak I know I suggested earlier to move this work into the `docParser`, but since it only keeps track of/iterates over the strings associated with the notices it doesn't know whether or not that return parameter is named via the AST and thus should enforce the error.  

Hope this solution isn't too hacky...

erak commented on libsolidity/interface/Natspec.cpp:
```diff
@@ -100,15 +100,52 @@ Json::Value Natspec::devDocumentation(ContractDefinition const& _contractDef)
 			if (!method.empty())
+			{
 				// add the function, only if we have any documentation to add
+				Json::Value ret(Json::objectValue);
```
Could use uniform initialization and maybe a more descriptive name e.g. `Json::Value jsonReturn{Json::objectValue}`

erak commented on libsolidity/interface/Natspec.cpp:
```diff
@@ -100,15 +100,52 @@ Json::Value Natspec::devDocumentation(ContractDefinition const& _contractDef)
+
+				if (!returnParams.empty())
+				{
+					unsigned int n = 0;
```
Since your incrementing `n` in the loop body in every case, you could remove this variable and use `i` for placeholder creation. Or is `returnDoc.first` returning a value > 0?

corydickson replied:
`i` is a map iterator; maybe I'm missing something on how to use that to access the vector for the return parameters. 😅 

Would changing the type for `n` to`size_t` help?

erak replied:
Oh, true. You're right. We need to iterate the map. I think the implementation is fine then and just needs some minor cleanup. One last proposal: maybe move the extraction into a function?
```
Json::Value Natspec::extractReturnParameterDocs(std::multimap<std::string, DocTag> const& _tags, FunctionDefinition const& _functionDef)
{
    if (!_functionDef.returnParameters().empty())
    {
        size_t index = 0;
        for (auto i = returnDoc.first; i != returnDoc.second; i++)
        {
            string paramName = _functionDef.returnParameters().at(n)->name();
            string content = i->second.content;
            ...
        }
    }
}
```
And then call it here:
```
if (!method.empty())
{
    auto returnDoc = fun->annotation().docTags.equal_range("return");
    
    Json::Value returnParameterDocs = extractReturnParameterDocs(returnDoc, *fun);
    
    if (!returnParameterDocs.empty())
	    method["return"] = returnParameterDocs;
    ....
}
```

erak commented on libsolidity/interface/Natspec.cpp:
```diff
@@ -100,15 +100,52 @@ Json::Value Natspec::devDocumentation(ContractDefinition const& _contractDef)
+
+						if (paramName.empty())
+						{
+							paramName = "_" + std::to_string(n+1);
```
Rethinking this again, I'd tend to start with `_0` as the first placeholder. That would probably be more inline with what people are expecting about array indices. I think you can just use `to_string(i)` then.

erak commented on libsolidity/interface/Natspec.cpp:
```diff
@@ -100,15 +100,52 @@ Json::Value Natspec::devDocumentation(ContractDefinition const& _contractDef)
+						string content = i->second.content;
+
+						if (paramName.empty())
+						{
```
`{` is not needed.

erak commented on libsolidity/interface/Natspec.cpp:
```diff
@@ -129,12 +166,6 @@ Json::Value Natspec::devDocumentation(std::multimap<std::string, DocTag> const&
 	if (!author.empty())
 		json["author"] = author;
 
-	// for constructors, the "return" node will never exist. invalid tags
```
Could you please also add a test case for this?

corydickson replied:
Does this one not provide proper coverage?

https://github.com/gh1dra/solidity/blob/29ec9445d859f425f97f54418912f29967a1a568/test/libsolidity/SolidityNatspecJSON.cpp#L831-L844

corydickson replied:
I'll add one for the new error introduced as well.

erak replied:
Yes, you're right. It's already partially covered. But a new one would be good.

erak commented:
@gh1dra Great, thanks! I left a few comments that should be easy to fix. And I also changed the base branch to `develop_060` since it's breaking. Could you please rebase your changes on top of it as well?
We need an entry in `solidity_repo/docs/060-breaking-changes.rst`  as well. You can run the HTML generation with `./scripts/docs.sh`.

erak commented on test/libsolidity/SolidityNatspecJSON.cpp:
```diff
@@ -408,13 +411,84 @@ BOOST_AUTO_TEST_CASE(dev_return_desc_after_nl)
+			/// @return d The result of the multiplication
+			/// @return f And cookies with nutella
+			function mul(uint a, uint second) public returns (uint d, uint f) {
+        uint mul = a * 7;
```
Indentation is off.

corydickson committed:
Implement @erak review notes by creating function and adding constructor test

Update 060 doc with natspec change

Add two more tests with mixed usage

Fix solc-js

fix changelog

corydickson commented:
@erak Thanks for the feedback! Let me know if you'd like me to squash the earlier commits.

erak commented on Changelog.md:
```diff
@@ -11,7 +11,19 @@ Breaking changes:
  * Standard JSON Interface: Add option to disable or choose hash method between IPFS and Swarm for the bytecode metadata.
  * Syntax: ``push(element)`` for dynamic storage arrays do not return the new length anymore.
  * Type checker: Resulting type of exponentiation is equal to the type of the base. Also allow signed types for the base.
+ * Natspec JSON Interface: Support multiple ``@return`` statements in dev documentation to behave like named parameters.
```
I think we need to be a little bit more specific and mention both, the fixed support and the breaking implication of multiple return statements in the dev docs.

```suggestion
 * Natspec JSON Interface: Support multiple ``@return`` statements in ``@dev`` documentation and force them to have define a name if parameter is named.
```

erak commented on docs/060-breaking-changes.rst:
```diff
@@ -54,6 +54,7 @@ New Features
 ============
 
  * The :ref:`try/catch statement <try-catch>` allows you to react on failed external calls.
+ * Natspec supports multiple return parameters in dev documentation, enforcing the same naming check as ``@param``.
```
Sorry, I haven't been specific enough earlier, but can you please also add an entry in the `How to update your code` section starting something like this:
```
* For every named return parameter in a function's ``@dev`` documentation define a ``@return`` 
  entry which contains the parameter's name as the first word. E.g. if you have function ``f()`` defined 
  like ``function f() public returns (uint value)`` and a ``@dev`` annotating it, document its return 
  parameters like so: ``@return value The return value.``.
```

erak commented on test/libsolidity/SolidityNatspecJSON.cpp:
```diff
@@ -408,13 +411,84 @@ BOOST_AUTO_TEST_CASE(dev_return_desc_after_nl)
+	"            \"a\": \"Documentation for the first parameter starts here. Since it's a really complicated parameter we need 2 lines\",\n"
+	"            \"second\": \"Documentation for the second parameter\"\n"
+	"        },\n"
+	"        \"return\": {\n"
```
Mmmh, now that I'm reading it again, I'd tend to use `returns`, as in `params` instead of `return`. It's also clearly a key for more that one return parameter then. Since this is breaking anyways, I'd propose to clean it up properly. @chriseth What's your take on this?

erak commented on Changelog.md:
```diff
@@ -11,7 +11,19 @@ Breaking changes:
+
+
+
+### 0.5.13 (unreleased)
```
There's an additional headline that slipped down there: https://github.com/ethereum/solidity/pull/7534/files#diff-9b8fab691c00b9e5380b19ce882f3271R44.

erak commented on libsolidity/interface/Natspec.cpp:
```diff
@@ -100,15 +100,53 @@ Json::Value Natspec::devDocumentation(ContractDefinition const& _contractDef)
+				solAssert(content.substr(0, nameEndPos) == paramName, "No return param name given: " + paramName);
+				content = content.substr(nameEndPos+1);
+			}
+		jsonReturn[paramName] = Json::Value(content);
```
Indentation is off here.

erak commented on test/externalTests/solc-js/DAO/ManagedAccount.sol:
```diff
@@ -1,4 +1,5 @@
 /*
+ * Natspec JSON Interface: Support multiple ``@return`` statements in dev documentation to behave like named parameters.
```
Is this the line from the Changelog?

corydickson replied:
Lol yes my bad vim buffer slip

erak commented on test/libsolidity/SolidityNatspecJSON.cpp:
```diff
@@ -365,7 +365,7 @@ BOOST_AUTO_TEST_CASE(dev_return)
 			/// Since it's a really complicated parameter we need 2 lines
 			/// @param second Documentation for the second parameter
-			/// @return The result of the multiplication
+			/// @return d The result of the multiplication
```
We'd probably need a test for unnamed return parameters.

erak commented on test/libsolidity/SolidityNatspecJSON.cpp:
```diff
@@ -408,13 +411,84 @@ BOOST_AUTO_TEST_CASE(dev_return_desc_after_nl)
+	"        },\n"
+	"        \"return\": {\n"
+	"            \"_0\": \"The result of the multiplication\",\n"
+	"            \"_1\": \"And cookies with nutella\"\n"
```
Maybe add another test for mixed unnamed / named?

corydickson replied:
All add two, the first to ensure that the counter is labeling unamed parameters correctly by having it in between two named parameters, and another that has named parameter right after an unamed one.

erak commented:
@gh1dra I've read through it again carefully and found some small issues. I think we can agree on https://github.com/ethereum/solidity/pull/7534#discussion_r339732570 quickly and incorporate the outcome. Otherwise your PR looks really good and I'm looking forward to get it merged.

erak commented:
@gh1dra Sorry, but it also has merge conflicts.

corydickson commented:
The branch should be rebased properly now, apologies for the confusion

erak commented on test/compilationTests/gnosis/MarketMakers/LMSRMarketMaker.sol:
```diff
@@ -128,7 +128,9 @@ contract LMSRMarketMaker is MarketMaker {
     /// @param funding Initial funding for market
     /// @param outcomeIndex Index of exponential term to extract (for use by marginal price function)
-    /// @return A result structure composed of the sum, the offset used, and the summand associated with the supplied index
+    /// @return sum of the outcomes
```
```suggestion
    /// @return sum Sum of the outcomes
```

erak replied:
Ok, that's probably nit-picking, but these could be changed to ``/// @return sum The Sum of outcomes`` etc.

erak commented on libsolidity/interface/Natspec.h:
```diff
@@ -60,6 +61,7 @@ class Natspec
 	/// @return      A JSON representation
 	///              of the contract's developer documentation
 	static Json::Value devDocumentation(std::multimap<std::string, DocTag> const& _tags);
+	static Json::Value extractReturnParameterDocs(std::multimap<std::string, DocTag> const& _tags, FunctionDefinition const& _functionDef);
```
This could use some more documentation, perhaps similar to the function ``devDocumentation`` above.

corydickson committed:
Add documentation for extractReturnParameterDocs

erak commented on Changelog.md:
```diff
@@ -12,6 +12,7 @@ Breaking changes:
  * Syntax: ``push(element)`` for dynamic storage arrays do not return the new length anymore.
  * Inline Assembly: Only strict inline assembly is allowed.
  * Type checker: Resulting type of exponentiation is equal to the type of the base. Also allow signed types for the base.
+ * Natspec JSON Interface: Support multiple ``@return`` statements in dev documentation to behave like parameters where the first word in the entry must contain the name.
```
```suggestion
 * Natspec JSON Interface: Properly support multiple ``@return`` statements in ``@dev`` documentation and enforce named return parameters to be mentioned documentation.
```

corydickson committed:
Update Changelog.md

author:@erak

Co-Authored-By: Erik K <erik@zirkular.io>

chriseth commented:
This looks good! I'll rebase and merge.
